web/cobrands/fixmystreet/assets.js

```diff
@@ -445,11 +442,79 @@ fixmystreet.assets = {
             fixmystreet.map.addControl(fixmystreet.assets.controls[i]);
             fixmystreet.assets.controls[i].activate();
         }
+    },
+
+    select_usrn: function(lonlat) {
+        var usrn_providers = fixmystreet.map.getLayersBy('fixmystreet', {
+            test: function(options) {
+                return options && options.usrn_field;
+            }
+        });
+        if (usrn_providers.length) {
+            var usrn_layer = usrn_providers[0];
+            var point = new OpenLayers.Geometry.Point(lonlat.lon, lonlat.lat);
+            var feature = usrn_layer.getFeatureAtPoint(point);
+            if (feature == null) {
+                // The click wasn't directly over a road, try and find one
+                // nearby
+                feature = usrn_layer.getNearestFeature(point, 10);
+            }
+            if (feature !== null) {
+                selected_usrn = feature.attributes[usrn_layer.fixmystreet.usrn_field];
+            } else {
+                selected_usrn = null;
+            }
+            fixmystreet.assets.update_usrn_field();
+        }
+    },
+
+    update_usrn_field: function() {
+        $("input[name=usrn]").val(selected_usrn);
     }
 };
 
 $(function() {
     fixmystreet.assets.init();
 });
 
+OpenLayers.Layer.Vector.prototype.getFeatureAtPoint = function(point) {
+    for (var i = 0; i < this.features.length; i++) {
+        var feature = this.features[i];
+        if (feature.geometry.containsPoint && feature.geometry.containsPoint(point)) {
+            return feature;
+        }
+    }
+    return null;
+};
+
+
+/*
+ * Returns this layer's feature that's closest to the given
+ * OpenLayers.Geometry.Point, as long as it's within <threshold> metres.
+ * Returns null if no feature meeting these criteria is found.
+ */
+OpenLayers.Layer.Vector.prototype.getNearestFeature = function(point, threshold) {
+    var nearest_feature = null;
+    var nearest_distance = null;
+    for (var i = 0; i < this.features.length; i++) {
+        var candidate = this.features[i];
+        var details = candidate.geometry.distanceTo(point, {details: true});
+        if (nearest_distance === null || details.distance < nearest_distance) {
+            // The units used for details.distance aren't metres, they're
+            // whatever the map projection uses. Convert to metres in order to
+            // draw a meaningful comparison to the threshold value.
+            var p1 = new OpenLayers.Geometry.Point(details.x0, details.y0);
+            var p2 = new OpenLayers.Geometry.Point(details.x1, details.y1);
+            var line = new OpenLayers.Geometry.LineString([p1, p2]);
+            var distance_m = line.getGeodesicLength(this.map.getProjectionObject());
+
+            if (distance_m <= threshold) {
+                nearest_feature = candidate;
+                nearest_distance = details.distance;
```

Comment:
To save (potentially?) on some processing, do you want to always set nearest_distance so that it can skip any further away, even if outside the threshold?

Reply:
D'oh. Good idea!